veniq/baselines/semi/clustering.py

```diff
@@ -0,0 +1,112 @@
+from collections import Counter
+from typing import List, Union, Dict
+from collections import OrderedDict
+
+from veniq.baselines.semi.extract_semantic import extract_method_statements_semantic
+from veniq.ast_framework import ASTNode, AST
+from veniq.baselines.semi.common_cli import common_cli
+
+
+def _check_is_common(
+    dict_file: Dict,
+    statement_1: Union[int, ASTNode],
+    statement_2: Union[int, ASTNode]
+) -> bool:
+    joined_names: Counter = Counter(dict_file[statement_1] + dict_file[statement_2])
+    duplicates = {element: count for element, count in joined_names.items() if count > 1}.keys()
+    return len(list(duplicates)) >= 1
+
+
+def _is_in_range(elem: int, values: List[int]) -> bool:
+    return elem >= values[0] and elem <= values[1]
+
+
+def _process_statement(
+    dict_file: Dict,
+    step: int
+) -> List[List[int]]:
+    clusters: List[List[int]] = []
+    list_statements = list(dict_file.keys())
+    for stat_1 in list_statements:
+        stat_1_line = stat_1 if isinstance(stat_1, int) else stat_1.line
+        for stat_2 in list_statements[:stat_1_line + step]:
+            stat_2_line = stat_2 if isinstance(stat_2, int) else stat_2.line
+            if stat_1_line < stat_2_line and _check_is_common(dict_file, stat_1, stat_2):
+                if len(clusters) != 0 and _is_in_range(stat_1_line, clusters[-1]):
+                    if not _is_in_range(stat_2_line, clusters[-1]):
+                        clusters[-1][1] = stat_2_line
+                else:
+                    clusters.append([stat_1_line, stat_2_line])
+    return clusters
+
+
+def SEMI(dict_file: Dict[ASTNode, List[str]]) -> List[List[int]]:
+    opportunities = []
+
+    first_statement_ = list(dict_file.keys())[0]
+    last_statement_ = list(dict_file.keys())[-1]
+    first_statement = first_statement_ if isinstance(first_statement_, int) else first_statement_.line
+    last_statement = last_statement_ if isinstance(last_statement_, int) else last_statement_.line
+    method_length = last_statement - first_statement
+
+    for step in range(1, method_length + 1):
+        clusters = _process_statement(dict_file, step)
+        opportunities += clusters
+    unique_oppo = [list(oppo) for oppo in set(map(tuple, opportunities))]
```

Comment:
1. why do you need a `set` operator? also `tuple` seems no need?
2. Line 55 looks redundant. Can we return `opportunities` instead?

Reply:
I think it is clear now from our talk